spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/SparkFileWriterFactory.java

```diff
@@ -340,4 +346,14 @@ SparkFileWriterFactory build() {
           writeProperties);
     }
   }
+
+  private static StructType calculateSparkType(StructType sparkType, Schema schema) {
+    if (sparkType != null) {
+      return sparkType;
+    } else if (schema != null) {
+      return SparkSchemaUtil.convert(schema);
+    } else {
+      return null;
```

Comment:
fail here?

Reply:
ok. I remember a previous discussion. for data schema, it should fail. for equality delete schema, it can be null. I am favoring two separate helper methods so that it does strict check for data schema.

Reply:
reopen this thread. I was saying null is not an option for data schema. we may need to helper methods for different behaviors for data and equality schemas.